src/Components/Web/src/Virtualization/VirtualizeDeferred.cs

```diff
@@ -0,0 +1,109 @@
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information. 
+
+using System;
+using System.Collections.Concurrent;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Components.Rendering;
+
+namespace Microsoft.AspNetCore.Components.Virtualization
+{
+    /// <summary>
+    /// Provides functionality for rendering a virtualized list from a deferred source.
+    /// </summary>
+    /// <typeparam name="TItem">The <c>context</c> type for the items being rendered.</typeparam>
+    public class VirtualizeDeferred<TItem> : VirtualizeBase<TItem>
+    {
+        private readonly ConcurrentQueue<TItem> _loadedItems = new ConcurrentQueue<TItem>();
+
+        private int _isFetchingItems;
+
+        private int _itemCount;
+
+        /// <summary>
+        /// Gets or sets the item template for the list.
+        /// </summary>
+        [Parameter]
+        public RenderFragment<TItem> Item { get; set; } = default!;
+
+        /// <summary>
+        /// Gets or sets the template for items that have not yet been loaded in memory.
+        /// </summary>
+        [Parameter]
+        public RenderFragment<int> Placeholder { get; set; } = default!;
+
+        /// <summary>
+        /// Gets or sets the function providing items to the list.
+        /// </summary>
+        [Parameter]
+        public ItemsProviderDelegate<TItem> ItemsProvider { get; set; } = default!;
+
+        /// <inheritdoc />
+        protected override int ItemCount => _itemCount;
+
+        /// <inheritdoc />
+        protected override void OnParametersSet()
+        {
+            if (Item == null)
+            {
+                throw new InvalidOperationException($"{GetType()} requires the template parameter '{nameof(Item)}' to be specified.");
+            }
+
+            if (Placeholder == null)
+            {
+                throw new InvalidOperationException($"{GetType()} requires the template parameter '{nameof(Placeholder)}' to be specified.");
+            }
+
+            if (ItemsProvider == null)
+            {
+                throw new InvalidOperationException($"{GetType()} requires a non-null parameter '{nameof(ItemsProvider)}'.");
+            }
+        }
+
+        /// <inheritdoc />
+        protected override void RenderItems(RenderTreeBuilder builder)
+        {
+            var itemsRendered = 0;
+
+            foreach (var item in _loadedItems.Skip(ItemsBefore).Take(ItemsVisible))
+            {
+                Item(item)(builder);
+                itemsRendered++;
+            }
+
+            for (; itemsRendered < ItemsVisible; itemsRendered++)
+            {
+                Placeholder(ItemsBefore + ItemsVisible + itemsRendered)(builder);
+            }
+
+            var itemsToFetch = ItemsBefore + VisibleItemCapacity - _loadedItems.Count;
+
+            if (itemsToFetch > 0)
+            {
+                _ = FetchItemsAsync(itemsToFetch);
```

Comment:
We did some work in Router where we cancel a previously executing operation before continuing. https://github.com/dotnet/aspnetcore/blob/master/src/Components/Components/src/Routing/Router.cs#L202. We should do something similar here. Right now, you could have concurrently executing fetches with no way of cancelling them.

Reply:
I believe the system I have in place involving `_isFetchingItems` prevents two fetches from happening at once. The reason I did it this way instead of cancelling the task is that if the user continuously scrolls through a list of loading items, there will be a multitude of tasks being run and almost instantly cancelled due to another task being run asking for a different range of items, none of them completing until the user stops scrolling and lets the last task finish. Currently, `FetchItemsAsync` no-ops if there is already a fetch in progress, so fetches won't be cancelling each other. When the current fetch completes, the component state is updated, and the next fetch can make a request for an item range that covers what each of the no-oped fetches would have collectively ask for.

Do you think this is an acceptable solution? What I don't have now is a way to bubble-up exceptions, but I've implemented a fix locally to support it which I plan to commit soon.

Reply:
I agree with your assessment. Unlike with navigation, we can assume that many or even most of the data requests would be discarded unless we do something to throttle it.

It sounds like what you have implemented already is a natural kind of backpressure whereby we don't issue the Nth request until the (N-1)th has completed. This has the nice property that the granularity of requests will scale according to the user's request times.

However this model does have the drawback that if just *one* of these requests hangs for a long time for some random reason, let's say for 30 seconds due to some 3G weirdness, then no subsequent requests will be made until that long period elapses, and the user will be left staring at a blank screen even if a subsequent request would have completed instantly.

There are other possible backpressure strategies that don't have this issue. Since different tradeoffs exist, I'd recommend we decouple the `<Virtualize>` component's data-requesting strategy from the underlying data provider's request-issuing-and-cancelling strategy. That way developer have the flexibility to wire up `<Virtualize>` to data providers that implement different strategies, plus it simplifies `<Virtualize>` as you won't have to deal with this here.

For example:

 * `<Virtualize>` could ask the `ItemsProvider` for data very eagerly, every time it sees any change to the scroll position that would bring new items into view, also immediately cancelling any previous request.
 * The underlying `ItemsProvider` can respond to this in different ways:
    * For providers that wrap an in-memory collection, they can respond instantly so there's never any lag in the UI
    * For providers that are making expensive network requests, they can:
      * ... have debounce strategy like "always wait for the previous request first, plus discard any requests that are superseded by a new queued one" like you have here. This is simple and would be good for fetching data from disk where you know the latency is bounded.
      * ... have a debounce strategy like "pause each outbound request by N milliseconds before we issue it, and discard if another request arrives while we were paused", so that requests never get queued behind their antecedents, but also never happen faster that one per N milliseconds [A]
      * ... have a more complicated debounce strategy like the above one plus "but if no requests are in flight then issue new ones immediately" to reduce latency in slow-scrolling cases, and maybe "if requests keep arriving faster than the N milliseconds timeout for more than M milliseconds, then issue the next one immediately" so the user doesn't have to fully stop scrolling to see a new batch arrive.

Basically, there's power and flexibility if we decouple the underlying `ItemProvider`'s strategy from `<Virtualize>`.

It's debatable whether we should *also* build in a basic default debounce mechanism into `<Virtualize>` such as control over the N in strategy [A] above, with a default of zero which skips that form of debouncing. That could help some customers who aren't comfortable implementing their own debounce mechanism. However I think we could do this additively in the future, so it doesn't need to be part of the initial implementation.